### PR TITLE
Debug draw

### PR DESCRIPTION
**Pull request by RobotCaleb**
image

image

### CODE REVIEW

File: main.cc [hunk 1]
```diff
@@ -430,6 +441,8 @@ void render() {
     camera_params.ptr->aspect = (float)wnd.width / wnd.height;
     camera_params.bind(0, frame);
 
+    ddRenderIfaceGL->mvpMatrix = camera_params.ptr->view_proj_matrix;
```
Comment (chrisforbes) [inline main.cc]: Don't read this back. It's quite possibly USWC, not cacheable.

File: nightmare.vcxproj [hunk 2]
```diff
@@ -189,7 +190,7 @@
     <ClInclude Include="src\component\gas_producer_component.h" />
     <ClInclude Include="src\component\light_component.h" />
     <ClInclude Include="src\component\physics_component.h" />
-    <ClInclude Include="src\component\power_component.h" />
+    <ClInclude Include="src\compebonent\power_component.h" />
```
Comment (chrisforbes) [inline nightmare.vcxproj]: This isn't right.

Also, missing filters. The files show up in random places in the project tree.

File: main.cc (hunk 1 above)
Comment (RobotCaleb) [reply main.cc]: alright, thanks

File: nightmare.vcxproj (hunk 2 above)
Comment (RobotCaleb) [reply nightmare.vcxproj]: yes, that's bad.
Okay, I'll not update the VS side except with VS.

**Comment by RobotCaleb** [comment]
fixt

File: src/utils/debugdraw_gl.cc [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+    }
+}
+
+static void compileShader(const GLuint shader)
+{
+    glCompileShader(shader);
+    checkGLError(__FILE__, __LINE__);
+
+    GLint status;
+    glGetShaderiv(shader, GL_COMPILE_STATUS, &status);
```
Comment (chrisforbes) [inline src/utils/debugdraw_gl.cc]: Use the existing shader support rather than adding another copy.

File: src/utils/debugdraw_gl.cc [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+        default                   : return "Unknown GL error";
+    } // switch (errorCode)
+}
+
+static void checkGLError(const char * file, const int line)
+{
+    GLenum err;
+    while ((err = glGetError()) != GL_NO_ERROR)
+    {
+        errorF("%s(%d) : GL_CORE_ERROR=0x%X - %s", file, line, err, errorToString(err));
```
Comment (chrisforbes) [inline src/utils/debugdraw_gl.cc]: We don't do this. KHR_debug announces our errors to us.

File: src/utils/debugdraw_gl.cc [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+// ========================================================
+
+const char * DDRenderInterfaceCoreGL::linePointVertShaderSrc = R"(
+    #version 150
+
+    in vec3 in_Position;
+    in vec4 in_ColorPointSize;
+
+    out vec4 v_Color;
+    uniform mat4 u_MvpMatrix;
```
Comment (chrisforbes) [inline src/utils/debugdraw_gl.cc]: Tweak these shaders to take the same per-camera uniform block everything else does, and a bunch of the other noise in this file goes away. (along with needing more copies of your view*proj matrix in main)

**Comment by chrisforbes** [comment]
Generally unhappy with the GL integration now that I've read it. Will fix.

**Comment by chrisforbes** [comment]
Fixing my grumbles in a followup PR.

**Pull request merged**